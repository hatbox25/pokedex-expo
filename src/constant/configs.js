export const BASE_API_URL = "https://pokeapi.co/api/v2/";
export const API_TIMEOUT = 15000; //milisecond
