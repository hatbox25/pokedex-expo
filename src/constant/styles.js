export default {
  primaryColor: "rgb(246,207,83)",
  secondaryColor: "rgb(24,26,31)",
  backgroundColor: "rgb(234,234,234)",
};
