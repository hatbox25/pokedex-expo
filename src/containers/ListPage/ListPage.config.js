import PropTypes from "prop-types";

const displayName = "ListPage";
const propTypes = {
  getPokemonList: PropTypes.func.isRequired,
  getPokemonByType: PropTypes.func.isRequired,
  pokemons: PropTypes.shape({
    data: PropTypes.array,
    loading: PropTypes.bool,
  }),
};
const defaultProps = {};

export default { displayName, propTypes, defaultProps };
