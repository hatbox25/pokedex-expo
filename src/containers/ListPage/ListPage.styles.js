import defaultStyle from "../../constant/styles";

export default {
  container: {
    flex: 1,
    alignItems: "center",
    backgroundColor: defaultStyle.backgroundColor,
  },

  itemRow: (index) => ({
    padding: 16,
    flexDirection: "row",
    justifyContent: "space-between",
    backgroundColor: index % 2 === 1 ? "#fff" : defaultStyle.backgroundColor,
  }),

  bottomSheetHeader: {
    backgroundColor: defaultStyle.secondaryColor,
    justifyContent: "center",
    alignItems: "center",
    height: 32,
    borderTopLeftRadius: 16,
    borderTopRightRadius: 16,
  },
  bottomSheetContent: {
    backgroundColor: defaultStyle.secondaryColor,
    height: 68,
  },

  optionRow: {
    flexDirection: "row",
    justifyContent: "space-between",
  },
  buttonOption: (selected) => ({
    flex: 1,
    backgroundColor: defaultStyle.primaryColor,
    color: defaultStyle.secondaryColor,
    margin: 10,
    padding: 8,
    borderRadius: 4,
    justifyContent: "center",
    alignItems: "center",
    borderWidth: selected ? 3 : 0,
    borderColor: defaultStyle.backgroundColor,
  }),
};
