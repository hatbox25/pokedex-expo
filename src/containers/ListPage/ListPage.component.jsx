import React, { PureComponent } from "react";
import { View, Text, FlatList, TouchableOpacity } from "react-native";
import BottomSheet from "reanimated-bottom-sheet";

import { capitalizeWord } from "../../utils/string.util";

import config from "./ListPage.config";
import styles from "./ListPage.styles";

import defaultStyle from "../../constant/styles";

const POKEMON_TYPE = [
  {
    id: "4",
    name: "Poison",
  },
  {
    id: "5",
    name: "Ground",
  },
  {
    id: "9",
    name: "Steel",
  },
];

class ListPage extends PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      bottomSheetOpen: false,
      selectedType: "",
    };

    this.bottomSheetRef = React.createRef();
  }
  componentDidMount() {
    // call api to get pokemon list in first time
    this._getPokemonList(true, false)();
  }

  _getPokemonList = (isRefresh, isAppend) => () => {
    if (!(isAppend && this.state.selectedType)) {
      // don't allow call api when endreact with selected filter
      const {
        pokemons: { data },
        getPokemonList,
      } = this.props;
      const params = {
        limit: 20,
        offset: !isRefresh ? data.length : 0,
      };
      this.setState({ selectedType: "" });
      getPokemonList(params, isAppend);
    }
  };

  _onClickPokemonItem = (url) => () => {
    const pokemonId = url.split("/")[6];
    this.props.navigation.push("Detail", { pokemonId });
  };

  renderItem = ({ item: { name, url }, index }) => {
    return (
      <TouchableOpacity
        style={styles.itemRow(index)}
        onPress={this._onClickPokemonItem(url)}
      >
        <Text style={{ color: defaultStyle.secondaryColor }}>
          {capitalizeWord(name)}
        </Text>
        <Text style={{ color: defaultStyle.secondaryColor }}>{">"}</Text>
      </TouchableOpacity>
    );
  };

  renderBottomSheetHeader = () => {
    const { selectedType, bottomSheetOpen } = this.state;
    let headerText = "Pull for Filter by Type";
    if (bottomSheetOpen) {
      headerText = "Select Pokemon Type";
    } else if (selectedType) {
      headerText = `Selected type: ${this.state.selectedType}, pull to change`;
    }
    return (
      <View style={styles.bottomSheetHeader}>
        <Text style={{ color: defaultStyle.primaryColor }}>{headerText}</Text>
      </View>
    );
  };

  _onClickFilterItem = (type) => () => {
    this.setState({ selectedType: type.name }, () => {
      this.props.getPokemonByType(type.id);
      this.bottomSheetRef.current.snapTo(0); // to close bottom sheet
    });
  };

  renderFilterOption = (type) => (
    <TouchableOpacity
      testID={`filter_type_${type.name}`}
      style={styles.buttonOption(this.state.selectedType === type.name)}
      onPressIn={this._onClickFilterItem(type)}
    >
      <Text>{type.name}</Text>
    </TouchableOpacity>
  );

  renderBottomSheetContent = () => (
    <View style={styles.bottomSheetContent}>
      <View style={styles.optionRow}>
        {POKEMON_TYPE.map(this.renderFilterOption)}
      </View>
    </View>
  );

  _setBottomSheet = (bottomSheetOpen) => () => {
    this.setState({ bottomSheetOpen });
  };

  render() {
    const { data, loading } = this.props.pokemons;
    return (
      <View style={styles.container}>
        <FlatList
          refreshing={loading}
          onRefresh={this._getPokemonList(true, false)}
          onEndReached={this._getPokemonList(false, true)}
          onEndReachedThreshold={0}
          style={{ width: "100%" }}
          data={data}
          renderItem={this.renderItem}
          keyExtractor={(item) => item.url}
        />
        <BottomSheet
          ref={this.bottomSheetRef}
          enabledManualSnapping
          snapPoints={[32, 100]}
          onOpenEnd={this._setBottomSheet(true)}
          onCloseEnd={this._setBottomSheet(false)}
          renderContent={this.renderBottomSheetContent}
          renderHeader={this.renderBottomSheetHeader}
        />
      </View>
    );
  }
}

ListPage.displayName = config.displayName;
ListPage.defaultProps = config.defaultProps;
ListPage.propTypes = config.propTypes;

export default ListPage;
