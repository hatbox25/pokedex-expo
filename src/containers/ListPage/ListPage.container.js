import { connect } from "react-redux";

import ListPage from "./ListPage.component";

import {
  getPokemonList,
  getPokemonByType,
} from "../../redux/actions/list.action";

const mapStateToProps = (state) => ({
  pokemons: state.list,
});
const mapDispatchToProps = { getPokemonList, getPokemonByType };

export default connect(mapStateToProps, mapDispatchToProps)(ListPage);
