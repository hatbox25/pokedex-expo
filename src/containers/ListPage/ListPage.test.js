import React from "react";
import { shallow } from "enzyme";
import thunk from "redux-thunk";
import configureStore from "redux-mock-store";

import ListPageContainer from "./ListPage.container";
import ListPage from "./ListPage.component";
import config from "./ListPage.config";

import { listPokomen } from "../../constant/fictures";

jest.mock("reanimated-bottom-sheet", () => "BottomSheet");

const mockStore = configureStore([thunk]);
const initialState = {};
const store = mockStore(initialState);

const props = {
  ...config.defaultProps,
  pokemons: {
    data: listPokomen,
    loading: false,
  },
  navigation: {
    push: jest.fn(),
  },
  getPokemonList: jest.fn(() => Promise.resolve()),
  getPokemonByType: jest.fn(() => Promise.resolve()),
};

let wrapper;
let instance;
beforeEach(() => {
  wrapper = shallow(<ListPage {...props} />);
  instance = wrapper.instance();
});

afterEach(() => {
  jest.clearAllMocks();
});

describe("ListPage Unit Test", () => {
  describe("_getPokemonList", () => {
    test("should call getPokemonList correctly on infinite list with selected filter", () => {
      wrapper.setState({ selectedType: "Poison" });
      instance._getPokemonList(false, true)();

      expect(props.getPokemonList).toBeCalled();
    });
  });

  describe("_onClickPokemonItem", () => {
    test("should navigate to pokemon detail with correct url", () => {
      instance._onClickPokemonItem("https://pokeapi.co/api/v2/pokemon/1/")();

      expect(props.navigation.push).toBeCalledWith("Detail", {
        pokemonId: "1",
      });
    });
  });

  describe("_onClickFilterItem", () => {
    test("should call api to get filtered result & close bottom sheet", () => {
      instance.bottomSheetRef = { current: { snapTo: jest.fn() } };

      instance._onClickFilterItem({ id: 4, name: "Poison" })();

      expect(props.getPokemonByType).toBeCalledWith(4);
    });
  });

  describe("_setBottomSheet", () => {
    test("should set bottomSheetOpen state correctly", () => {
      instance._setBottomSheet(true)();

      expect(wrapper.state("bottomSheetOpen")).toEqual(true);
    });
  });
});

describe("ListPage Snap Test", () => {
  test("should render correctly", () => {
    expect(shallow(<ListPageContainer store={store} />)).toMatchSnapshot();
    expect(wrapper).toMatchSnapshot();
  });

  test("should render item correctly", () => {
    const result = instance.renderItem({
      item: { name: "Bulbasaur", url: "UrlBulbasaur" },
      index: 1,
    });
    expect(result).toMatchSnapshot();
  });

  test("should render bottom sheet header", () => {
    const result = instance.renderBottomSheetHeader();
    expect(result).toMatchSnapshot();
  });

  test("should render bottom sheet header when open", () => {
    wrapper.setState({ bottomSheetOpen: true });

    const result = instance.renderBottomSheetHeader();
    expect(result).toMatchSnapshot();
  });

  test("should render bottom sheet header when filter selected", () => {
    wrapper.setState({ selectedType: "Poison" });

    const result = instance.renderBottomSheetHeader();
    expect(result).toMatchSnapshot();
  });

  test("should render bottom sheet content", () => {
    const result = instance.renderBottomSheetContent();
    expect(result).toMatchSnapshot();
  });
});
