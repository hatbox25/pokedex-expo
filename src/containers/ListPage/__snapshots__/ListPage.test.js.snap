// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListPage Snap Test should render bottom sheet content 1`] = `
<View
  style={
    Object {
      "backgroundColor": "rgb(24,26,31)",
      "height": 68,
    }
  }
>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <TouchableOpacity
      activeOpacity={0.2}
      focusedOpacity={0.7}
      onPressIn={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "rgb(246,207,83)",
          "borderColor": "rgb(234,234,234)",
          "borderRadius": 4,
          "borderWidth": 0,
          "color": "rgb(24,26,31)",
          "flex": 1,
          "justifyContent": "center",
          "margin": 10,
          "padding": 8,
        }
      }
      testID="filter_type_Poison"
    >
      <Text>
        Poison
      </Text>
    </TouchableOpacity>
    <TouchableOpacity
      activeOpacity={0.2}
      focusedOpacity={0.7}
      onPressIn={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "rgb(246,207,83)",
          "borderColor": "rgb(234,234,234)",
          "borderRadius": 4,
          "borderWidth": 0,
          "color": "rgb(24,26,31)",
          "flex": 1,
          "justifyContent": "center",
          "margin": 10,
          "padding": 8,
        }
      }
      testID="filter_type_Ground"
    >
      <Text>
        Ground
      </Text>
    </TouchableOpacity>
    <TouchableOpacity
      activeOpacity={0.2}
      focusedOpacity={0.7}
      onPressIn={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "rgb(246,207,83)",
          "borderColor": "rgb(234,234,234)",
          "borderRadius": 4,
          "borderWidth": 0,
          "color": "rgb(24,26,31)",
          "flex": 1,
          "justifyContent": "center",
          "margin": 10,
          "padding": 8,
        }
      }
      testID="filter_type_Steel"
    >
      <Text>
        Steel
      </Text>
    </TouchableOpacity>
  </View>
</View>
`;

exports[`ListPage Snap Test should render bottom sheet header 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "rgb(24,26,31)",
      "borderTopLeftRadius": 16,
      "borderTopRightRadius": 16,
      "height": 32,
      "justifyContent": "center",
    }
  }
>
  <Text
    style={
      Object {
        "color": "rgb(246,207,83)",
      }
    }
  >
    Pull for Filter by Type
  </Text>
</View>
`;

exports[`ListPage Snap Test should render bottom sheet header when filter selected 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "rgb(24,26,31)",
      "borderTopLeftRadius": 16,
      "borderTopRightRadius": 16,
      "height": 32,
      "justifyContent": "center",
    }
  }
>
  <Text
    style={
      Object {
        "color": "rgb(246,207,83)",
      }
    }
  >
    Selected type: Poison, pull to change
  </Text>
</View>
`;

exports[`ListPage Snap Test should render bottom sheet header when open 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "rgb(24,26,31)",
      "borderTopLeftRadius": 16,
      "borderTopRightRadius": 16,
      "height": 32,
      "justifyContent": "center",
    }
  }
>
  <Text
    style={
      Object {
        "color": "rgb(246,207,83)",
      }
    }
  >
    Select Pokemon Type
  </Text>
</View>
`;

exports[`ListPage Snap Test should render correctly 1`] = `
<ContextProvider
  value={null}
>
  <ListPage
    getPokemonByType={[Function]}
    getPokemonList={[Function]}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  />
</ContextProvider>
`;

exports[`ListPage Snap Test should render correctly 2`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "rgb(234,234,234)",
      "flex": 1,
    }
  }
>
  <FlatList
    data={
      Array [
        Object {
          "name": "bulbasaur",
          "url": "https://pokeapi.co/api/v2/pokemon/1/",
        },
        Object {
          "name": "ivysaur",
          "url": "https://pokeapi.co/api/v2/pokemon/2/",
        },
        Object {
          "name": "venusaur",
          "url": "https://pokeapi.co/api/v2/pokemon/3/",
        },
        Object {
          "name": "charmander",
          "url": "https://pokeapi.co/api/v2/pokemon/4/",
        },
        Object {
          "name": "charmeleon",
          "url": "https://pokeapi.co/api/v2/pokemon/5/",
        },
        Object {
          "name": "charizard",
          "url": "https://pokeapi.co/api/v2/pokemon/6/",
        },
        Object {
          "name": "squirtle",
          "url": "https://pokeapi.co/api/v2/pokemon/7/",
        },
        Object {
          "name": "wartortle",
          "url": "https://pokeapi.co/api/v2/pokemon/8/",
        },
        Object {
          "name": "blastoise",
          "url": "https://pokeapi.co/api/v2/pokemon/9/",
        },
        Object {
          "name": "caterpie",
          "url": "https://pokeapi.co/api/v2/pokemon/10/",
        },
        Object {
          "name": "metapod",
          "url": "https://pokeapi.co/api/v2/pokemon/11/",
        },
        Object {
          "name": "butterfree",
          "url": "https://pokeapi.co/api/v2/pokemon/12/",
        },
        Object {
          "name": "weedle",
          "url": "https://pokeapi.co/api/v2/pokemon/13/",
        },
        Object {
          "name": "kakuna",
          "url": "https://pokeapi.co/api/v2/pokemon/14/",
        },
        Object {
          "name": "beedrill",
          "url": "https://pokeapi.co/api/v2/pokemon/15/",
        },
        Object {
          "name": "pidgey",
          "url": "https://pokeapi.co/api/v2/pokemon/16/",
        },
        Object {
          "name": "pidgeotto",
          "url": "https://pokeapi.co/api/v2/pokemon/17/",
        },
        Object {
          "name": "pidgeot",
          "url": "https://pokeapi.co/api/v2/pokemon/18/",
        },
        Object {
          "name": "rattata",
          "url": "https://pokeapi.co/api/v2/pokemon/19/",
        },
        Object {
          "name": "raticate",
          "url": "https://pokeapi.co/api/v2/pokemon/20/",
        },
      ]
    }
    disableVirtualization={true}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReached={[Function]}
    onEndReachedThreshold={0}
    onRefresh={[Function]}
    refreshing={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    style={
      Object {
        "width": "100%",
      }
    }
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
  <BottomSheet
    enabledManualSnapping={true}
    onCloseEnd={[Function]}
    onOpenEnd={[Function]}
    renderContent={[Function]}
    renderHeader={[Function]}
    snapPoints={
      Array [
        32,
        100,
      ]
    }
  />
</View>
`;

exports[`ListPage Snap Test should render item correctly 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
  focusedOpacity={0.7}
  onPress={[Function]}
  style={
    Object {
      "backgroundColor": "#fff",
      "flexDirection": "row",
      "justifyContent": "space-between",
      "padding": 16,
    }
  }
>
  <Text
    style={
      Object {
        "color": "rgb(24,26,31)",
      }
    }
  >
    Bulbasaur
  </Text>
  <Text
    style={
      Object {
        "color": "rgb(24,26,31)",
      }
    }
  >
    &gt;
  </Text>
</TouchableOpacity>
`;
